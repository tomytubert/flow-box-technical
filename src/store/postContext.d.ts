import { Response } from '../interfaces/types';
interface PostsContextType {
    data: Response | null;
    error: any;
    loading: boolean;
}
export declare function PostsContextProvider({ children, }: {
    children: React.ReactNode;
}): import("react/jsx-runtime").JSX.Element;
export declare const usePosts: () => PostsContextType;
export {};
