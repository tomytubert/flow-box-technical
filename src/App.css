* {
	scrollbar-width: none;
}
.capitalize-first-letter::first-letter {
	text-transform: capitalize;
}

@keyframes fadeIn {
	0% {
		opacity: 0;
	}
	100% {
		opacity: 1;
	}
}


.fade-in {
	animation-duration: 2s;
	animation-fill-mode: both;
	animation-name: fadeIn;
}
