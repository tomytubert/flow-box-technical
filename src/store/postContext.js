import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext } from 'react';
import useFetch from '../hooks/useFetch';
var PostsContext = createContext(null);
export function PostsContextProvider(_a) {
    var children = _a.children;
    var _b = useFetch("https://api.unsplash.com/search/photos?client_id=".concat(process.env.UNSPLASH_ACCESS_KEY, "&query=dogs&per_page=15&page=1")), data = _b.data, error = _b.error, loading = _b.loading;
    return (_jsx(PostsContext.Provider, { value: { data: data, loading: loading, error: error }, children: children }));
}
export var usePosts = function () {
    var context = useContext(PostsContext);
    if (!context) {
        throw new Error('usePosts must be used within a PostsContextProvider');
    }
    return context;
};
