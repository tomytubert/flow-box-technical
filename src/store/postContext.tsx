import { createContext, useContext } from 'react'
import useFetch from '../hooks/useFetch'
import { Response } from '../interfaces/types'

interface PostsContextType {
    data: Response | null
    error: any
    loading: boolean
}

const PostsContext = createContext<PostsContextType | null>(null)

export function PostsContextProvider({
	children,
}: {
	children: React.ReactNode
}) {
	const { data, error, loading } = useFetch(
		`https://api.unsplash.com/search/photos?client_id=${process.env.UNSPLASH_ACCESS_KEY}&query=dogs&per_page=15&page=1`
	)

	return (
		<PostsContext.Provider value={{ data, loading, error }}>
			{children}
		</PostsContext.Provider>
	)
}

export const usePosts = () => {
    const context = useContext(PostsContext)
    if (!context) {
        throw new Error('usePosts must be used within a PostsContextProvider')
    }
    return context
}
