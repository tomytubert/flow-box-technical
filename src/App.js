import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import './App.css';
import { AppBar, Box, Button, CircularProgress, Container, FormControl, InputLabel, MenuItem, Select, Typography, } from '@mui/material';
import { usePosts } from './store/postContext';
import { CardGrid } from './components/card-grid/CardGrid';
import { useState } from 'react';
import { Carousel } from './components/carousel/Carousel';
import { ImageGrid } from './components/grid/ImageGrid';
import { ImageList } from './components/list/ImageList';
var views = {
    carousel: 'CAROUSEL',
    card: 'CARD',
    grid: 'GRID',
    list: 'LIST',
};
export var App = function () {
    var _a = usePosts(), data = _a.data, loading = _a.loading, error = _a.error;
    var _b = useState(views.list), selectedView = _b[0], setSelectedView = _b[1];
    if (loading)
        return (_jsx(Container, { sx: {
                minHeight: '90dvh',
                justifyContent: 'center',
                alignItems: 'center',
                display: 'flex',
            }, children: _jsx(CircularProgress, { color: 'primary' }) }));
    if (error || !data)
        return (_jsx(Container, { sx: {
                minHeight: '90dvh',
                justifyContent: 'center',
                alignItems: 'center',
                display: 'flex',
            }, children: _jsxs(Box, { children: [_jsx(Typography, { variant: 'h6', children: "No data available, please try again" }), _jsx(Button, { size: 'medium', variant: 'contained', sx: { mt: 2 }, onClick: function () { return window.location.reload(); }, children: "Try again" })] }) }));
    var handleChange = function (e) {
        setSelectedView(e.target.value);
    };
    return (_jsxs(_Fragment, { children: [_jsx(AppBar, { position: 'static', sx: { p: 2, backgroundColor: 'primary.light' }, children: _jsxs(Box, { display: 'flex', justifyContent: 'space-between', alignItems: 'center', gap: 2, children: [_jsx(Typography, { variant: 'h6', children: "Your audience content" }), _jsxs(FormControl, { size: 'small', sx: { minWidth: 120 }, children: [_jsx(InputLabel, { id: 'demo-simple-select-label', sx: { color: 'common.white' }, children: "View" }), _jsxs(Select, { labelId: 'demo-simple-select-label', id: 'demo-simple-select', value: selectedView, label: 'View', onChange: handleChange, sx: { borderRadius: '16px', color: 'common.white' }, children: [_jsx(MenuItem, { value: views.carousel, children: "Carousel" }), _jsx(MenuItem, { value: views.card, children: "Card" }), _jsx(MenuItem, { value: views.grid, children: "Grid" }), _jsx(MenuItem, { value: views.list, children: "List" })] })] })] }) }), _jsxs(Container, { maxWidth: 'md', sx: { marginTop: '2rem', display: 'flex', placeContent: 'center' }, children: [selectedView === views.carousel && _jsx(Carousel, { className: "fade-in", posts: data.results }), selectedView === views.card && _jsx(CardGrid, { className: "fade-in", posts: data.results }), selectedView === views.grid && _jsx(ImageGrid, { className: "fade-in", posts: data.results }), selectedView === views.list && _jsx(ImageList, { className: "fade-in", posts: data.results })] })] }));
};
export default App;
