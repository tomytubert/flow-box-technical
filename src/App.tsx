import './App.css'
import {
	AppBar,
	Box,
	Container,
	FormControl,
	InputLabel,
	MenuItem,
	Select,
	SelectChangeEvent,
	Typography,
} from '@mui/material'
import { usePosts } from './store/postContext'
import { CardGrid } from './componentes/card-grid/CardGrid'
import { useState } from 'react'
import { Carousel } from './componentes/carousel/Carousel'
import { ImageGrid } from './componentes/grid/ImageGrid'

const views = {
	carousel: 'CAROUSEL',
	card: 'CARD',
	grid: 'GRID',
	list: 'LIST',
}

function App() {
	const { data, loading, error } = usePosts()
	const [selectedView, setSelectedView] = useState(views.grid)

	if (loading) return <p>Loading...</p>
	if (error || !data) return <p>Error</p>

	const handleChange = (e: SelectChangeEvent) => {
		setSelectedView(e.target.value)
	}

	return (
		<>
			<AppBar position='static' sx={{ p: 2, backgroundColor: 'primary.light' }}>
				<Box
					display='flex'
					justifyContent='space-between'
					alignItems='center'
					gap={4}
				>
					<Typography variant='h6'>Your audience content</Typography>
					<FormControl size='small' sx={{ minWidth: 120 }}>
						<InputLabel id='demo-simple-select-label'>View</InputLabel>
						<Select
							labelId='demo-simple-select-label'
							id='demo-simple-select'
							value={selectedView}
							label='View'
							onChange={handleChange}
						>
							<MenuItem value={views.carousel}>Carousel</MenuItem>
							<MenuItem value={views.card}>Card</MenuItem>
							<MenuItem value={views.grid}>Grid</MenuItem>
							<MenuItem value={views.list}>List</MenuItem>
						</Select>
					</FormControl>
				</Box>
			</AppBar>
			<Container
				maxWidth='md'
				sx={{ marginTop: '2rem', display: 'flex', placeContent: 'center' }}
			>
				{selectedView === views.carousel && <Carousel posts={data.results} />}
				{selectedView === views.card && <CardGrid posts={data.results} />}
				{selectedView === views.grid && <ImageGrid posts={data.results} />}
				{selectedView === views.list && <p>List</p>}
			</Container>
		</>
	)
}

export default App
