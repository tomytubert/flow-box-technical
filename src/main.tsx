import React from 'react'
import ReactDOM from 'react-dom/client'
import App from './App.tsx'
import './index.css'
import { CssBaseline, ThemeProvider, createTheme } from '@mui/material'
import Inter from './assets/fonts/Inter-VariableFont_opsz-wght.ttf'
import { TopBar } from './components/ui/topbar/TopBar.tsx'
import { PostsContextProvider } from './store/postContext.tsx'

const theme = createTheme({
	typography: {
		fontFamily: ['Inter'].join(','),
	},
	palette: {
		primary: {
			main: '#7F3DF3',
			light: '#9863F5',
			dark: '#582AAA',
			contrastText: '#fff',
		},
	},
	components: {
		MuiCssBaseline: {
			styleOverrides: `
        @font-face {
          font-family: 'Inter';
          font-style: normal;
          font-display: swap;
          font-weight: 400;
          src: local('Inter'), local('Inter-Regular'), url(${Inter}) format('truetype');
          unicodeRange: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF;
        }
      `,
		},
		MuiOutlinedInput: {
			styleOverrides: {
				root: {
					'& .MuiOutlinedInput-notchedOutline': {
						borderColor: '#fff',
					},
					'&.Mui-focused .MuiOutlinedInput-notchedOutline': {
						borderColor: '#582AAA',
					},
					'&:hover .MuiOutlinedInput-notchedOutline': {
						borderColor: '#582AAA',
					},
				},
			},
		},
		MuiSelect: {
			styleOverrides: {
				icon: {
					color: '#582AAA', // Change this to your desired color
				},
			},
		},
		MuiSvgIcon: {
			styleOverrides: {
				root: {
					'&.MuiSelect-icon': {
						color: '#fff', // Change this to your desired color
					},
				},
			},
		},
		MuiCard: {
            styleOverrides: {
                root: {
                    borderRadius: '16px', // Custom border radius
                    boxShadow: '0 4px 8px rgba(0, 0, 0, 0.1)', // Custom box shadow
                },
            },
        },
	},
})

ReactDOM.createRoot(document.getElementById('root')!).render(
	<React.StrictMode>
		<PostsContextProvider>
			<ThemeProvider theme={theme}>
				<CssBaseline />
				<TopBar />
				<App />
			</ThemeProvider>
		</PostsContextProvider>
	</React.StrictMode>
)
