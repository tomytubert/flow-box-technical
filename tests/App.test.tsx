import React from 'react'
import { App } from '../src/App'
import { fireEvent, render, screen } from '@testing-library/react'
import { usePosts } from '../src/store/postContext'

jest.mock('../src/store/postContext')

jest.mock('swiper/react', () => ({
	Swiper: ({ children }: any) => children,
	SwiperSlide: ({ children }: any) => children,
}))

jest.mock('swiper/modules', () => ({
	Navigation: jest.fn(),
	Pagination: jest.fn(),
}))

jest.mock('swiper/css', () => jest.fn())
jest.mock('swiper/css/pagination', () => jest.fn())

describe('Testing <App />', () => {
	test('should match the snapshot when loading is true', () => {
		;(usePosts as jest.Mock).mockReturnValue({
			data: null,
			loading: true,
			error: null,
		})
		const { container } = render(<App />)
		expect(container).toMatchSnapshot()
	})

	test('should match the snapshot when data is loaded', () => {
		;(usePosts as jest.Mock).mockReturnValue({
			data: { total: 10, total_pages: 5, results: [] },
			loading: false,
			error: null,
		})
		const { container } = render(<App />)
		expect(container).toMatchSnapshot()
	})

	test('should match the snapshot when there is an error', () => {
		;(usePosts as jest.Mock).mockReturnValue({
			data: null,
			loading: false,
			error: 'Error fetching posts',
		})
		const { container } = render(<App />)
		expect(container).toMatchSnapshot()
	})

	test('should change view when button is clicked', () => {
	(usePosts as jest.Mock).mockReturnValue({
			data: {
				total: 10,
				total_pages: 5,
				results: [],
			},
			loading: false,
			error: null,
		})
		render(<App />)
		const input = screen.getByDisplayValue('LIST') as HTMLInputElement
		fireEvent.click(input, { target: { value: 'GRID' } })
        expect(input.value).toBe('GRID')    
	})
})
