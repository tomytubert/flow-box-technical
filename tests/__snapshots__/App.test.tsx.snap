// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <App /> should match the snapshot when data is loaded 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-k3sv75-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiBox-root css-1o9f1gx"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        Your audience content
      </h6>
      <div
        class="MuiFormControl-root css-pe0pe4-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1xcs70u-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="demo-simple-select-label"
        >
          View
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-10o05uz-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-controls=":r0:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="demo-simple-select-label demo-simple-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="demo-simple-select"
            role="combobox"
            tabindex="0"
          >
            List
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="LIST"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                View
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd css-bvfh9d-MuiContainer-root"
  >
    <div
      class="fade-in MuiBox-root css-1l0qm2u"
    />
  </div>
</div>
`;

exports[`Testing <App /> should match the snapshot when loading is true 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-6350ea-MuiContainer-root"
  >
    <span
      class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`Testing <App /> should match the snapshot when there is an error 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-6350ea-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        No data available, please try again
      </h6>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-lll1vm-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Try again
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
